packages/lambda/src/lib/store/index.js

```diff
@@ -106,3 +92,47 @@ export class ClaimStorage {
     }))
   }
 }
+
+/**
+ * @param {import('@web3-storage/content-claims/server/api').Claim} claim
+ * @param {import('./s3-bucket').S3Bucket} s3
+ **/
+async function storeClaimBytes ({ claim, bytes }, { bucketName, s3Client }) {
+  const cidstr = claim.toString(base32)
+  const key = `${cidstr}/${cidstr}.car`
+  return retry(() => {
+    const cmd = new PutObjectCommand({
+      Bucket: bucketName,
+      Key: key,
+      Body: bytes
+    })
+    return s3Client.send(cmd)
+  }, {
+    minTimeout: 100,
+    onFailedAttempt: err => console.warn(`failed S3 put for: ${cidstr}`, err)
+  })
+}
+
+/**
+ * @param {import('@web3-storage/content-claims/server/api').Claim} claim
+ * @param {import('./dynamo-table').DynamoTable} dynamo
+ */
+async function upsertClaim ({ claim, content, expiration }, { tableName, dynamoClient }) {
+  const hasExpiration = expiration && isFinite(expiration)
+  const mh = base58btc.encode(content.bytes)
+  return retry(() => {
+    const cmd = new UpdateItemCommand({
+      TableName: tableName,
+      Key: marshall({
+        claim: claim.toString(),
+        content: mh
+      }),
+      ExpressionAttributeValues: marshall({ ':ex': hasExpiration ? expiration : 0 }),
+      UpdateExpression: 'SET expiration=:ex'
+    })
+    return dynamoClient.send(cmd)
+  }, {
+    minTimeout: 100,
+    onFailedAttempt: err => console.warn(`failed DynamoDB update for: ${mh}`, err)
```

Comment:
Yeah we should probably log the full primary key, something like?:

```suggestion
    onFailedAttempt: err => console.warn(`failed DynamoDB update for content: ${mh} claim: ${claim.toString(base32)}`, err)
```